src/backend/distributed/relay/relay_event_utility.c

```diff
@@ -536,6 +547,70 @@ AppendShardIdToConstraintName(AlterTableCmd *command, uint64 shardId)
 }
 
 
+static bool
+ExtendRowReferencesWalker(Node *node, RowRefWalkerState *state)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, IndexElem))
+	{
+		IndexElem *indexElem = (IndexElem *) node;
+
+		return raw_expression_tree_walker(indexElem->expr, ExtendRowReferencesWalker,
+										  state);
+	}
+	else if (IsA(node, ColumnRef))
+	{
+		ColumnRef *columnRef = (ColumnRef *) node;
+		ListCell *fieldsCell;
+
+		/*
+		 * Append the shardId to any ColumnRef String values that are
+		 * equal to the relationName.  These are actually ROW(relname)
+		 * references.
+		 */
+		foreach(fieldsCell, columnRef->fields)
```

Comment:
What if a relation, schema, and catalog (database) all share the same name? `foo.foo.foo.*` is allowed, but this code could transform it to `foo_123.foo_123.foo_123.*`, which would be bad.

[This comment](https://github.com/postgres/postgres/blob/REL9_5_4/src/backend/parser/parse_expr.c#L517-L539) enumerates _exactly_ all the syntaxes allowed in this parse location. A safe thing to do here would be to check that the last element in the `fields` list is a `A_Star` type and if it is, transform the second-to-last element to contain the shard identifier (which should be a string; maybe throw an `Assert` in).

[This piece of code](https://github.com/postgres/postgres/blob/REL9_5_4/src/backend/parser/parse_target.c#L134-L151) looks like it does something similar (checking for a `ColumnRef` that ends in a `A_Star`)… can probably drop it in here with code to check the penultimate element for a relation name match.